### PR TITLE
Switch to Renovate from Dependabot for remaining dependencies

### PR DESCRIPTION
**Pull request by timja**
See https://github.com/jenkins-infra/helpdesk/issues/4173

We were already using renovate for JavaScript dependencies, this switches us to using it for the remaining ones.
Primary motivator is buggy handling of SNAPSHOTS in dependabot.

### Testing done

Tested on my org fork (my personal fork was getting auth failures for some reason):
https://github.com/timja-org/jenkins/pulls?q=sort%3Aupdated-desc+is%3Apr+is%3Aopen

image

I downgraded remoting on my fork and verified it got upgraded correctly: https://github.com/timja-org/jenkins/pull/6

Full result of detected packages / package managers:
https://developer.mend.io/github/timja-org/jenkins

### Proposed changelog entries

N/A

### Proposed upgrade guidelines

N/A

```[tasklist]
### Submitter checklist
- [x] The Jira issue, if it exists, is well-described.
- [x] The changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developers, depending on the change) and are in the imperative mood (see [examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)). Fill in the **Proposed upgrade guidelines** section only if there are breaking changes or changes that may require extra steps from users during upgrade.
- [x] There is automated testing or an explanation as to why this change has no tests.
- [x] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadocs, as appropriate.
- [x] New deprecations are annotated with `@Deprecated(since = "TODO")` or `@Deprecated(forRemoval = true, since = "TODO")`, if applicable.
- [x] New or substantially changed JavaScript is not defined inline and does not call `eval` to ease future introduction of Content Security Policy (CSP) directives (see [documentation](https://www.jenkins.io/doc/developer/security/csp/)).
- [x] For dependency updates, there are links to external changelogs and, if possible, full differentials.
- [x] For new APIs and extension points, there is a link to at least one consumer.
```

### Desired reviewers

N/A

Before the changes are marked as `ready-for-merge`:

```[tasklist]
### Maintainer checklist
- [ ] There are at least two (2) approvals for the pull request and no outstanding requests for change.
- [ ] Conversations in the pull request are over, or it is explicit that a reviewer is not blocking the change.
- [x] Changelog entries in the pull request title and/or **Proposed changelog entries** are accurate, human-readable, and in the imperative mood.
- [x] Proper changelog labels are set so that the changelog can be generated automatically.
- [x] If the change needs additional upgrade steps from users, the `upgrade-guide-needed` label is set and there is a **Proposed upgrade guidelines** section in the pull request title (see [example](https://github.com/jenkinsci/jenkins/pull/4387)).
- [x] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).
```

### CODE REVIEW

File: .github/dependabot.yml
Comment (timja) [inline .github/dependabot.yml:68]: I don't think this ever worked, see https://github.com/jenkinsci/jenkins/pulls?q=sort%3Aupdated-desc+is%3Apr+is%3Aclosed+author%3Aapp%2Fdependabot+label%3Ainto-lts

so I haven't kept it for going forward

**Review by MarkEWaite — APPROVED** [review]
The proposed list of updates seems very reasonable to me.

The `args4j` 2.37 release that is included in the list of proposed upgrades was released in March 2024.  Good that renovate detected it.  I've not done any investigation to see if the upgrade is safe, since it will need evaluation in remoting as well as Jenkins core.

**Comment by basil** [comment]
Is this able to upgrade detached plugins like Dependabot?

**Comment by timja** [comment]
> Is this able to upgrade detached plugins like Dependabot?

It's upgrading `war/pom.xml` but not `test/pom.xml` for some reason:

https://github.com/timja-org/jenkins/pull/8

https://github.com/timja-org/jenkins/blob/b673f0d8108d2d03b5bb34d536060dc6741b107b/test/pom.xml#L353

Looking in to it

**Comment by timja** [comment]
> Is this able to upgrade detached plugins like Dependabot?

Yes, all working now that `test` isn't ignored: https://github.com/timja-org/jenkins/pull/8

**Comment by timja** [comment]
It is now trying to upgrade `old-remoting-for-test` and `remoting.minimum.supported.version` I'll see if I can stop that

**Review by NotMyFault — APPROVED** [review]

**Comment by timja** [comment]
> It is now trying to upgrade `old-remoting-for-test` and `remoting.minimum.supported.version` I'll see if I can stop that

@viceice is it possible to stop it updating those two intentionally old deps?

https://github.com/timja-org/jenkins/pull/10
https://github.com/timja-org/jenkins/pull/9

**Comment by viceice** [comment]
yes, with a package rule and an allowed version filter. you can also disable that package or require dashboard approval.

**Comment by timja** [comment]
> yes, with a package rule and an allowed version filter. you can also disable that package or require dashboard approval.

Hmm we still want the regular dependency one to work, https://github.com/timja-org/jenkins/pull/6
Just not the 'minimum version' or the 'old version to test'.

Would that still work?

I couldn't see a way to filter out a dependency from certain paths

Dashboard approval could maybe work

**Comment by viceice** [comment]
> > yes, with a package rule and an allowed version filter. you can also disable that package or require dashboard approval.
> 
> Hmm we still want the regular dependency one to work, https://github.com/timja-org/jenkins/pull/6
> Just not the 'minimum version' or the 'old version to test'.
> 
> Would that still work?
> 
> I couldn't see a way to filter out a dependency from certain paths
> 
> Dashboard approval could maybe work

this can do it:
https://docs.renovatebot.com/configuration-options/#matchfilenames

**Comment by timja** [comment]
This not right? https://github.com/timja-org/jenkins/blob/37f95213954136508d215a4972d701bae1689a8b/.github/renovate.json#L71

**Comment by viceice** [comment]
> This not right? https://github.com/timja-org/jenkins/blob/37f95213954136508d215a4972d701bae1689a8b/.github/renovate.json#L71

no. you're missing an action. that rule matches the file and excludes that dep. but you don't say what renovate should do. so add `enabled:false` to disable or allowed version to restrict only by version.

**Comment by timja** [comment]
Thanks that fixed one of them.

Surprisingly it didn't close this one though: https://github.com/timja-org/jenkins/pull/9

(It wasn't picked up when the entire file was ignored before adjusting ignore patterns)

Any idea?

**Comment by viceice** [comment]
I'll try to have a look again tomorrow, currently on holiday 🙃

**Comment by timja** [comment]
@viceice you have a chance to check it?

**Comment by viceice** [comment]
> @viceice you have a chance to check it?

not yet, sorry. feel free to open a support discussion on the renovate repo. maybe one of my co-maintainer will have more time to look at it. I'm back on track on Monday.

**Comment by timja** [comment]
Hey @viceice any thoughts?

**Comment by timja** [comment]
raised config help at https://github.com/renovatebot/renovate/discussions/30555

**Comment by timja** [comment]
All working now, @basil / others any concerns?

File: .github/renovate.json
Comment (viceice) [inline .github/renovate.json:72]: this rule doesn't match `pom.xml`, so renovate still suggests an update for `org.jenkins-ci.main:remoting` inside the root `pom.xml`

**Comment by viceice** [comment]
> Thanks that fixed one of them.
> 
> Surprisingly it didn't close this one though: [timja-org#9](https://github.com/timja-org/jenkins/pull/9)
> 
> (It wasn't picked up when the entire file was ignored before adjusting ignore patterns)
> 
> Any idea?

See my code comment for the solution

https://github.com/jenkinsci/jenkins/pull/9459/files/5895e0ebe23dea21644e42d3a7c32e4f63295615#diff-f82f7d36a61d22f640c25bdb8b0435bf9cb38679d99d5f460753fa668ee7cdb3R72

File: .github/renovate.json
Comment (timja) [reply .github/renovate.json:72]: yeah there's one `org.jenkins-ci.main:remoting` that we want updated, the `remoting.version` one which is the latest version.

We don't want the 'old version for testing', and the 'minimum supported version' updated. Which is all working with this setup now by moving it to `.mvn/maven.config`

File: .github/dependabot.yml
Comment (basil) [reply .github/dependabot.yml:68]: > I don't think this ever worked

Sad, given that it cost me so much to get that change accepted. I still think this is something that we want to automated, and we essentially do it manually for every LTS release, including recent ones.

File: .mvn/maven.config
Comment (basil) [inline .mvn/maven.config:3]: This workaround doesn't seem maintainable to me. I would be OK with less invasive workarounds, like not having a Maven property and duplicating the version number in the three places it's consumed (if that fixes the problem). Otherwise, download https://hub.docker.com/r/renovate/renovate, set [up logging](https://docs.renovatebot.com/config-overview/#logging-variables), run `renovate --platform=local`, and go to town debugging—it's all open-source JavaScript code, so it should be doable to add proper support for this use case, or at least to add enough print statements to find a cleaner workaround. :smile:

Comment (timja) [reply .mvn/maven.config:3]: why is it not maintainable to you? if someone is going to update the remoting minimum version it will be a conscious action and they will search the property find it and update.

Comment (basil) [reply .mvn/maven.config:3]: Using a system property rather than a Maven property unnecessarily increases visibility from the Maven API to the Java Platform API, and a wider surface area increases the likelihood something will be misused. In addition we've had problems in the past with IDEs not parsing `maven.config`; even if they've been resolved, this still seems like the wrong layer to define this. I've seen reviews on `apache/maven` where the Maven developers are pushing back against the use of the Java Platform APIs to get Maven properties, and there is a large category of Maven issues related to variable interpolation, so I've developed a gut feeling that a lot of subtle things can go wrong here with regard to Maven, hence my preference for a simpler workaround (such as duplicating the property, if that works) which might at least have a more understandable failure mode.

File: .github/renovate.json
Comment (viceice) [inline .github/renovate.json:33]:
```suggestion
      "matchPackageNames": ["jakarta.servlet:jakarta.servlet-api"]
```
why regex?

Comment (timja) [reply .github/renovate.json:33]: the package validator did that when adapting from old syntax to new

File: .github/dependabot.yml
Comment (timja) [reply .github/dependabot.yml:68]: Restored in https://github.com/jenkinsci/jenkins/pull/9459/commits/74eee38e3652690d828ff081db16f78abd5b1bfd

had to debug a fair but and used:
```
RENOVATE_CONFIG_FILE=.github/renovate.json renovate --token=$GITHUB_TOKEN --schedule=""  --require-config=ignored --dry-run=full timja-org/jenkins
```

Comment (timja) [reply .github/dependabot.yml:68]: Here are the LTS PRs that would be created: https://github.com/timja-org/jenkins/pulls?q=sort%3Aupdated-desc+is%3Apr+is%3Aopen+label%3Ainto-lts

(some remoting ones would disappear by backporting this PR to LTS)

**Comment by timja** [comment]
I hit a bunch of weird behaviour when I added `baseBranches` in, `matchPackageNames` wildcarding and regex didn't seem to work.

I've managed to fix almost everything although this rule isn't working right now:
```json
    {
      "matchPackageNames": ["net.jcip:jcip-annotations"],
      "matchDatasources": ["maven"],
      "enabled": false,
      "description": "maven-metadata.xml is missing for this really old package which is required by renovate"
    }
```

its still complaining about not being able to resolve it and if I just drop it to master or stable branch only then its fine with that rule (not a blocker anyway it just fixes a warning).

**Review by viceice — APPROVED** [review]
looks good

**Review by basil — APPROVED** [review]
Thanks for the PR!

**Comment by timja** [comment]
/label ready-for-merge

----

This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

**Comment by basil** [comment]
There is a problem visible in #9529: Renovate is trying to update the GitHub Action to the latest commit, not the latest tagged commit. We want the same behavior as Dependabot, where updates are only offered once a commit is tagged.

**Comment by basil** [comment]
There is a problem visible in https://github.com/jenkinsci/jenkins/pulls: Renovate is trying to backport all Java library updates, not just security updates. We want the same behavior we were trying to enable in Dependabot, where [only security updates are backported](https://docs.renovatebot.com/presets-security/#securityonly-security-updates).

**Comment by viceice** [comment]
add a version tag comment, that is understood by dependabot and renovate. then they follow and update that tag.

eg `uses: org/repo@<sha> # v1.2.3`

https://docs.renovatebot.com/modules/manager/github-actions/

**Comment by basil** [comment]
> There is a problem visible in https://github.com/jenkinsci/jenkins/pulls: Renovate is trying to backport all Java library updates, not just security updates. We want the same behavior we were trying to enable in Dependabot, where [only security updates are backported](https://docs.renovatebot.com/presets-security/#securityonly-security-updates).

@timja Can you try re-adding the security updates with `osvVulnerabilityAlerts` using `matchBaseBranches` inside `packageRules` (per https://github.com/renovatebot/renovate/discussions/20542#discussioncomment-10268896)? `vulnerabilityAlerts` seems to be a non-starter, as it uses the GitHub Dependabot alerts (which only work on the default branch).

**Comment by basil** [comment]
@timja https://github.com/jenkinsci/jenkins-test-harness/releases/tag/2254.vcff7a_d4969e5 was released 7 hours ago, but Renovate does not seem to be offering it to Jenkins core, even after manually triggering Renovate. Dependabot is offering the update in other repositories, such as https://github.com/jenkinsci/plugin-pom/pull/984.

**Comment by timja** [comment]
> @timja [jenkinsci/jenkins-test-harness@`2254.vcff7a_d4969e5` (release)](https://github.com/jenkinsci/jenkins-test-harness/releases/tag/2254.vcff7a_d4969e5) was released 7 hours ago, but Renovate does not seem to be offering it to Jenkins core, even after manually triggering Renovate. Dependabot is offering the update in other repositories, such as [jenkinsci/plugin-pom#984](https://github.com/jenkinsci/plugin-pom/pull/984).

resolved

> @timja Can you try re-adding the security updates with osvVulnerabilityAlerts using matchBaseBranches inside packageRules (per https://github.com/renovatebot/renovate/discussions/20542#discussioncomment-10268896)? vulnerabilityAlerts seems to be a non-starter, as it uses the GitHub Dependabot alerts (which only work on the default branch).

I'm trying to get it working, haven't managed yet though.

**Pull request merged**
